diff -wurpN --no-dereference -X diffgen.ignore test-tree/drivers/firmware/efi/libstub/efi-stub.c milky/drivers/firmware/efi/libstub/efi-stub.c
--- test-tree/drivers/firmware/efi/libstub/efi-stub.c	2024-06-01 13:40:57.584170632 -0400
+++ milky/drivers/firmware/efi/libstub/efi-stub.c	2024-05-21 05:22:27.000000000 -0400
@@ -38,7 +38,7 @@
 #define EFI_RT_VIRTUAL_BASE	SZ_512M
 #define EFI_RT_VIRTUAL_SIZE	SZ_512M
 
-#ifdef CONFIG_ARM64
+#if defined(CONFIG_ARM64) || defined(CONFIG_RISCV)
 # define EFI_RT_VIRTUAL_LIMIT	DEFAULT_MAP_WINDOW_64
 #else
 # define EFI_RT_VIRTUAL_LIMIT	TASK_SIZE
