diff -wurpN --no-dereference -X diffgen.ignore test-tree/include/linux/mtd/spi-nor.h milky/include/linux/mtd/spi-nor.h
--- test-tree/include/linux/mtd/spi-nor.h	2024-06-01 13:47:58.906874556 -0400
+++ milky/include/linux/mtd/spi-nor.h	2024-05-21 05:22:27.000000000 -0400
@@ -26,6 +26,7 @@
 #define SPINOR_OP_WRSR		0x01	/* Write status register 1 byte */
 #define SPINOR_OP_RDSR2		0x3f	/* Read status register 2 */
 #define SPINOR_OP_WRSR2		0x3e	/* Write status register 2 */
+#define SPINOR_OP_WRSR2_JY	0x31    /* Write status register 2 like juyang */
 #define SPINOR_OP_READ		0x03	/* Read data bytes (low frequency) */
 #define SPINOR_OP_READ_FAST	0x0b	/* Read data bytes (high frequency) */
 #define SPINOR_OP_READ_1_1_2	0x3b	/* Read data bytes (Dual Output SPI) */
