diff -wurpN --no-dereference -X diffgen.ignore test-tree/include/uapi/linux/elf.h milky/include/uapi/linux/elf.h
--- test-tree/include/uapi/linux/elf.h	2024-06-01 13:47:59.402873044 -0400
+++ milky/include/uapi/linux/elf.h	2024-05-21 05:22:27.000000000 -0400
@@ -431,6 +431,7 @@ typedef struct elf64_shdr {
 #define NT_MIPS_DSP	0x800		/* MIPS DSP ASE registers */
 #define NT_MIPS_FP_MODE	0x801		/* MIPS floating-point mode */
 #define NT_MIPS_MSA	0x802		/* MIPS SIMD registers */
+#define NT_RISCV_VECTOR	0x900		/* RISC-V vector registers */
 
 /* Note types with note name "GNU" */
 #define NT_GNU_PROPERTY_TYPE_0	5
