diff -wurpN --no-dereference -X diffgen.ignore test-tree/include/trace/events/kmem.h milky/include/trace/events/kmem.h
--- test-tree/include/trace/events/kmem.h	2024-06-01 13:47:59.390873080 -0400
+++ milky/include/trace/events/kmem.h	2024-05-21 05:22:27.000000000 -0400
@@ -363,6 +363,48 @@ TRACE_EVENT(rss_stat,
 		__entry->member,
 		__entry->size)
 	);
+
+TRACE_EVENT(ion_heap_shrink,
+
+	TP_PROTO(const char *heap_name,
+		 size_t len,
+		 long total_allocated),
+
+	TP_ARGS(heap_name, len, total_allocated),
+
+	TP_STRUCT__entry(
+		__field(const char *, heap_name)
+		__field(size_t, len)
+		__field(long, total_allocated)
+	),
+
+	TP_fast_assign(
+		__entry->heap_name = heap_name;
+		__entry->len = len;
+		__entry->total_allocated = total_allocated;
+	),
+
+	TP_printk("heap_name=%s, len=%zu, total_allocated=%ld",
+		  __entry->heap_name, __entry->len, __entry->total_allocated)
+	);
+
+TRACE_EVENT(ion_heap_grow,
+		TP_PROTO(const char *heap_name, size_t len, long total_allocated),
+
+		TP_ARGS(heap_name, len, total_allocated),
+
+		TP_STRUCT__entry(__field(const char *, heap_name)
+				     __field(size_t, len)
+					     __field(long, total_allocated)),
+
+		TP_fast_assign(__entry->heap_name = heap_name; __entry->len = len;
+			   __entry->total_allocated = total_allocated;),
+
+		TP_printk("heap_name=%s, len=%zu, total_allocated=%ld",
+		      __entry->heap_name, __entry->len,
+		      __entry->total_allocated)
+		);
+
 #endif /* _TRACE_KMEM_H */
 
 /* This part must be outside protection */
