diff -wurpN --no-dereference -X diffgen.ignore test-tree/kernel/compat.c milky/kernel/compat.c
--- test-tree/kernel/compat.c	2024-06-01 13:47:59.446872909 -0400
+++ milky/kernel/compat.c	2024-05-21 05:22:27.000000000 -0400
@@ -269,24 +269,3 @@ get_compat_sigset(sigset_t *set, const c
 	return 0;
 }
 EXPORT_SYMBOL_GPL(get_compat_sigset);
-
-/*
- * Allocate user-space memory for the duration of a single system call,
- * in order to marshall parameters inside a compat thunk.
- */
-void __user *compat_alloc_user_space(unsigned long len)
-{
-	void __user *ptr;
-
-	/* If len would occupy more than half of the entire compat space... */
-	if (unlikely(len > (((compat_uptr_t)~0) >> 1)))
-		return NULL;
-
-	ptr = arch_compat_alloc_user_space(len);
-
-	if (unlikely(!access_ok(ptr, len)))
-		return NULL;
-
-	return ptr;
-}
-EXPORT_SYMBOL_GPL(compat_alloc_user_space);
