diff -wurpN --no-dereference -X diffgen.ignore test-tree/include/linux/uaccess.h milky/include/linux/uaccess.h
--- test-tree/include/linux/uaccess.h	2024-06-01 13:47:59.334873251 -0400
+++ milky/include/linux/uaccess.h	2024-05-21 05:22:27.000000000 -0400
@@ -11,6 +11,7 @@
 #include <asm/uaccess.h>
 
 #ifdef CONFIG_SET_FS
+
 /*
  * Force the uaccess routines to be wired up for actual userspace access,
  * overriding any possible set_fs(KERNEL_DS) still lingering around.  Undone
@@ -200,16 +201,6 @@ copy_to_user(void __user *to, const void
 		n = _copy_to_user(to, from, n);
 	return n;
 }
-#ifdef CONFIG_COMPAT
-static __always_inline unsigned long __must_check
-copy_in_user(void __user *to, const void __user *from, unsigned long n)
-{
-	might_fault();
-	if (access_ok(to, n) && access_ok(from, n))
-		n = raw_copy_in_user(to, from, n);
-	return n;
-}
-#endif
 
 #ifndef copy_mc_to_kernel
 /*
