diff -wurpN --no-dereference -X diffgen.ignore test-tree/include/uapi/linux/media-bus-format.h milky/include/uapi/linux/media-bus-format.h
--- test-tree/include/uapi/linux/media-bus-format.h	2024-06-01 13:47:59.410873020 -0400
+++ milky/include/uapi/linux/media-bus-format.h	2024-05-21 05:22:27.000000000 -0400
@@ -153,6 +153,16 @@
 /* S5C73M3 sensor specific interleaved UYVY and JPEG */
 #define MEDIA_BUS_FMT_S5C_UYVY_JPEG_1X8		0x5001
 
+/* CVITEK RGBIR specific formats */
+#define MEDIA_BUS_FMT_CVI_BAYER_GRGBI_1x12	0x5101
+#define MEDIA_BUS_FMT_CVI_BAYER_RGBGI_1x12	0x5102
+#define MEDIA_BUS_FMT_CVI_BAYER_GBGRI_1x12	0x5103
+#define MEDIA_BUS_FMT_CVI_BAYER_BGRGI_1x12	0x5104
+#define MEDIA_BUS_FMT_CVI_BAYER_IGRGB_1x12	0x5105
+#define MEDIA_BUS_FMT_CVI_BAYER_IRGBG_1x12	0x5106
+#define MEDIA_BUS_FMT_CVI_BAYER_IBGRG_1x12	0x5107
+#define MEDIA_BUS_FMT_CVI_BAYER_IGBGR_1x12	0x5108
+
 /* HSV - next is	0x6002 */
 #define MEDIA_BUS_FMT_AHSV8888_1X32		0x6001
 
