diff -wurpN --no-dereference -X diffgen.ignore test-tree/kernel/reboot.c milky/kernel/reboot.c
--- test-tree/kernel/reboot.c	2024-06-01 13:47:59.454872885 -0400
+++ milky/kernel/reboot.c	2024-05-21 05:22:27.000000000 -0400
@@ -416,7 +416,7 @@ void ctrl_alt_del(void)
 		kill_cad_pid(SIGINT, 1);
 }
 
-char poweroff_cmd[POWEROFF_CMD_PATH_LEN] = "/sbin/poweroff";
+char poweroff_cmd[POWEROFF_CMD_PATH_LEN] = "/sbin/poweroff -f";
 static const char reboot_cmd[] = "/sbin/reboot";
 
 static int run_cmd(const char *cmd)
